### PR TITLE
implement alembic to handle database migrations

### PR DESCRIPTION
**Pull request by RustyBower**
### Description
Implement Alembic to handle current and future database migrations

### Checklist
- [X] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [X] I can and do license this contribution under the EFLv2
- [X] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [X] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by RustyBower** [comment]
I do need to figure out how to automatically run Alembic for new versions of Sopel 🤔

**Comment by dgw** [comment]
That'd be cool. I'm not clear on how this works with packaged installs vs. being installed from Git, because the version appears based on commit SHAs, but also haven't read this whole patch yet (nor read Alembic docs).

Maybe later when I'm not out and on a phone. 😁

**Comment by Exirel** [comment]
As said on IRC, I'm sorry I've to say no, because I've been there, I've done the work on previous project so I know what it means. However when we want migrations in Sopel:

* it needs to be a sopel command line, allowing to select a different configuration file
* it needs to allow us to generate new migrations that can be packaged
* it needs to allow Sopel's owner to run migrations that was packaged
* it needs to allow plugin's author to generate new migrations for their plugin (in that case, maybe only packaged version?)
* it needs to collect plugin's migrations to run them alongside Sopel's own migrations

Basically, we don't need the default behavior of Alembic, which is to run for a project you 100% control, we need a system closer to what Django's ORM support: generate migrations for your project, package it, then collect all migrations from the project and its installed plugin to run them. And trust me, I know it's not easy, I had to done it once a few years ago and it wasn't fun at all.

As a side note, the migration needs to have a complete downgrade handler, this one is incomplete.

**Comment by RustyBower** [comment]
There's definitely a way to package this within the application itself to migrate on boot / startup, since this is functionality that Flask-Migrate has, so it must be possible.

**Comment by Exirel** [comment]
I know there is a way: I've done it. I also know this PR is not the way.

**Comment by RustyBower** [comment]
I'm open to other suggestions

**Comment by dgw** [comment]
This seems to have well and truly stalled. Since it didn't seem to be the right approach anyway, let's take any future discussion back to #2293. I'll copy over @Exirel's list of requirements.
